### PR TITLE
Improve handling of published/unpublished data in downloads

### PR DESCRIPTION
**Pull request by munnsmunns**
Adds a checkbox to allow team members or higher to choose to include unpublished data in downloads.
Makes the permissions to download unpublished data **team member or higher** consistently across all downloads.
image

### CODE REVIEW

File: hawc/apps/assessment/templates/assessment/assessment_downloads.html [hunk 1]
```diff
@@ -184,6 +184,90 @@ <h3 class="card-header">Ecological</h3>
...
+        <span class="d-inline-block" style="width:200px">Published & Unpublished:</span>
+        <a class="btn btn-sm btn-success mx-1" href="{{the_url}}?unpublished=true&format=xlsx">xlsx</a>
+        <a class="btn btn-sm btn-success mx-1" href="{{the_url}}?unpublished=true&format=tsv">tsv</a>
+        <a class="btn btn-sm btn-success mx-1" href="{{the_url}}?unpublished=true&format=csv">csv</a>
+        <a class="btn btn-sm btn-success mx-1" href="{{the_url}}?unpublished=true&format=json">json</a>
+        <a class="btn btn-sm btn-success mx-1" href="{{the_url}}?unpublished=true&format=html">html</a>
+        <p class="text-muted mt-1">Help text</p>
+
+        <p><b>Animal bioassy export</b></p>
+        {% url 'animal:api:assessment-full-export' assessment.pk as the_url %}
```
Comment (shapiromatron) [inline hawc/apps/assessment/templates/assessment/assessment_downloads.html:214]: Can you make a component for this little collection of spans and links?

```
{% include 'assessment/components/foo.html' with url=the_url %}
```

Something like that, but with a better name :)

File: hawc/apps/assessment/templates/assessment/assessment_downloads.html [hunk 2]
```diff
@@ -184,6 +184,90 @@ <h3 class="card-header">Ecological</h3>
     </div>
   </div>
 </div>
+
+
+<p class='form-text text-muted mb-3'>
```
Comment (shapiromatron) [inline hawc/apps/assessment/templates/assessment/assessment_downloads.html:189]: only all added content if team member or higher?

File: hawc/apps/assessment/templates/assessment/assessment_downloads.html [hunk 3]
```diff
@@ -184,6 +184,90 @@ <h3 class="card-header">Ecological</h3>
...
+  Assessment downloads availalbe for team members or higher.
+</p>
+<div class="row">
+<div class="col-xl-6">
+    <h3>Data extractions</h3>
+    <div class="bg-light border shadow p-3">
+        <p><b>Animal bioassy export</b></p>
+        {% url 'animal:api:assessment-full-export' assessment.pk as the_url %}
```
Comment (shapiromatron) [inline hawc/apps/assessment/templates/assessment/assessment_downloads.html:197]: Can you make a component for this little collection of spans and links?

```
{% include 'assessment/components/foo.html' with url=the_url %}
```

Something like that, but with a better name :)

File: hawc/apps/common/api/utils.py [hunk 4]
```diff
@@ -15,7 +15,7 @@ def get_published_only(assessment: Assessment, request: Request) -> bool:
         request (Request): a DRF request instance
     """
     unpublished_requested = request.query_params.get("unpublished", "").lower() == "true"
-    can_edit = assessment.user_can_edit_object(request.user)
+    can_edit = assessment.user_is_team_member_or_higher(request.user)
```
Comment (shapiromatron) [inline hawc/apps/common/api/utils.py:18]: good catch

File: tests/hawc/apps/epi/test_api.py [hunk 5]
```diff
@@ -14,9 +14,11 @@
 
 @pytest.mark.django_db
 class TestEpiAssessmentViewSet:
-    def _test_flat_export(self, rewrite_data_files: bool, fn: str, url: str):
+    def _test_flat_export(
```
Comment (shapiromatron) [inline tests/hawc/apps/epi/test_api.py:17]: this is a nice refactoring

**Review by shapiromatron — COMMENTED** [review]
looking good - a few minor updates to the templates, but code looks good

**Comment by munnsmunns** [comment]
@shapiromatron should any of the lit, risk of bias, in vitro, or epi-meta exports have the unpublished filter added?

**Comment by shapiromatron** [comment]
> @shapiromatron should any of the lit, risk of bias, in vitro, or epi-meta exports have the unpublished filter added?

Yes, we'll need it for invitro, epimeta, and riskofbias.  We wont for lit.  I think invitro and epimeta may not be to difficult and will be similar to epi/epiv2, but riskofbias may be really different. I can help if needed.

File: hawc/apps/riskofbias/api.py [hunk 6]
```diff
@@ -67,12 +72,17 @@ def full_export(self, request, pk):
         Get all risk of bias/study evaluations for an assessment, including individual reviews.
         """
         self.get_object()
+        ser = ExportQuerySerializer(data=request.query_params)
+        ser.is_valid(raise_exception=True)
+        published_only = get_published_only(self.assessment, request)
         rob_name = self.assessment.get_rob_name_display().lower()
         exporter = exports.RiskOfBiasCompleteFlat(
```
Comment (shapiromatron) [inline hawc/apps/riskofbias/api.py:79]: The published_only variable should be passed in here and handled in the class init.  It should be available in the kwargs dictionary on the instance.

File: hawc/apps/riskofbias/api.py [hunk 7]
```diff
@@ -47,12 +48,16 @@ def export(self, request, pk):
...
+        ser = ExportQuerySerializer(data=request.query_params)
+        ser.is_valid(raise_exception=True)
+        published_only = get_published_only(self.assessment, request)
         rob_name = self.assessment.get_rob_name_display().lower()
         exporter = exports.RiskOfBiasFlat(
             self.get_queryset().none(),
             filename=f"{self.assessment}-{rob_name}",
             assessment_id=self.assessment.id,
         )
+        exporter.published_only = published_only
```
Comment (shapiromatron) [inline hawc/apps/riskofbias/api.py:60]: Same comment as below; pass in via the init

**Review by shapiromatron — CHANGES_REQUESTED** [review]
Looking great, a few mostly minor changes requested; it should be ready after those are in. I really like the UI change you made; it's much simpler and more intuitive than what I did.

**Review by shapiromatron — APPROVED** [review]
Looks great! I made two very minor edits; if you're ok w/ changes, ok to merge.

* 4e2d5b5f whitespace edits (12 seconds ago)
* 018010a9 only add the unpublished flag if the checkbox exists (57 minutes ago)

**Pull request merged**